genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcJobTerminationListenerServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.execution.services.impl.grpc;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.netflix.genie.agent.execution.ExecutionContext;
+import com.netflix.genie.agent.execution.services.JobTerminationListenerService;
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Implementation of the {@link JobTerminationListenerService}.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Lazy
+@Service
+@Slf4j
+public class GRpcJobTerminationListenerServiceImpl implements JobTerminationListenerService {
+
+    private final JobTerminationServiceGrpc.JobTerminationServiceFutureStub client;
+    private final ExecutionContext executionContext;
+
+    /**
+     * Constructor.
+     *
+     * @param client           The gRPC client to use to call the server.
+     * @param executionContext The execution context for the agent
+     */
+    public GRpcJobTerminationListenerServiceImpl(
+        final JobTerminationServiceGrpc.JobTerminationServiceFutureStub client,
+        final ExecutionContext executionContext
+    ) {
+        this.client = client;
+        this.executionContext = executionContext;
+    }
+
+
+    @Override
+    public void registerForRemoteKillNotification(@NotNull final Runnable onKillCallback) {
+        final String jobId = executionContext.getClaimedJobId();
+
+        Futures.addCallback(
+            this.client
+                .registerForKillNotification(
+                    JobKillListenerRequest.newBuilder()
+                        .setJobId(jobId)
+                        .build()
+                ),
+            new JobKillFutureCallback(onKillCallback)
+        );
+    }
+
+    private void handleErrorListeningForKillNotification(final Runnable onSuccessCallback) {
+        registerForRemoteKillNotification(onSuccessCallback);
+    }
+
+    private void handleKillNotification(final Runnable callback) {
+        callback.run();
+    }
+
+    private class JobKillFutureCallback implements FutureCallback<JobKillListenerResponse> {
+
+        private final Runnable onSuccessCallback;
+
+        JobKillFutureCallback(
+            final Runnable onSuccessCallback
+        ) {
+            this.onSuccessCallback = onSuccessCallback;
+        }
+
+        @Override
+        public void onSuccess(@Nullable final JobKillListenerResponse result) {
+            final String jobId = executionContext.getClaimedJobId();
```

Comment:
Rather than this why do you save the entire `ExecutionContext` as a member rather than just the job id and save this call to the getter?

Reply:
sure